### ISSUE TITLE
variable leaking within partial (scoping problem)

### ISSUE THREAD

**Issue by dadaxr** [statamic/cms#7587]
### Bug description

Variable defined inside a partial leak outside (ie : they are accessible in the main template after the partial tag has been processed). I've checked on another site of mine, in v3.3.12 (using new parsing engine too) and the behavior is not the same, ie : variables defined inside a partial does not leak outside.

Also, a side effect of this is, variables passed to partial (as parameter) could be ignored / overrided because of the leaks. See bellow 2nd example

### How to reproduce

Regarding the variable leak : 

in a partial file (_test.antlers.html)
```
{{ definedLocally = "i'm defined in the partial" }}
```

in the main template : 
```
{{ definedLocally }} {{# output nothing #}}
{{ partial:partials/_test }}
{{ definedLocally }} {{# output "i'm defined in the partial" instead of nothing #}}
```

Regarding the "annoying" side effect mentioned above : 

in a partial file (_test.antlers.html)
```
{{ classes = classes ?? 'defaultClass' }}
<div class="{{ classes }}">foo</div>
```

in the main template : 
```
{{ partial:partials/_test classes="" }} {{# output : <div class="defaultClass">foo</div> this is correct #}}
{{ partial:partials/_test classes="overridedClass" }} {{# output : <div class="defaultClass">foo</div> also,  this is incorrect #}}
{{ classes }}  {{# output : defaultClass , instead of nothing, because of the leak, this is incorrect #}}
```

### Environment

```yaml
php please support:details

Environment
Application Name: Statamic leporc
Laravel Version: 9.52.0
PHP Version: 8.1.14
Composer Version: 2.5.1
Environment: local
Debug Mode: ENABLED
URL: leporc.localhost
Maintenance Mode: OFF

Cache
Config: NOT CACHED
Events: NOT CACHED
Routes: NOT CACHED
Views: CACHED

Drivers
Broadcasting: log
Cache: statamic
Database: mysql
Logs: stack / single
Mail: smtp
Queue: sync
Session: file

Statamic
Addons: 1
Antlers: runtime
Stache Watcher: Disabled
Static Caching: Disabled
Version: 3.4.4 PRO

Statamic Addons
statamic/seo-pro: 4.0.1
```

### Installation

Fresh statamic/statamic site via CLI

### Antlers Parser

runtime (new)

**Comment by JohnathonKoster** [comment]
I am unable to reproduce this using only the samples provided on 3.4.4. Is there more of the templates leading up to the issue you could share, or possibly a sample repository recreating the issue?

**Comment by dadaxr** [comment]
Ok I refined the problem after more tests : it seems the leak comes from using a partial parameter name which is the same as an existing variable name in outer scope.

to reproduce : 

in a "home" template : 
```
{{ param1 = "defined as var in home" }}
{{ partial:partials/_test param1="defined as param in home" param2="defined as param in home" }}
home after <br>
param1 : {{ param1 }} <br> {{# output : "modified in partial" : INCORRECT should be "defined as var in home"  #}}
param2 : {{ param2 }} <br> {{# output : "" : CORRECT  ! #}}
param3 : {{ param3 }} <br> {{# output : "" CORRECT !  #}}
``` 

in a partial "_test" template : 
```
<hr>
param1 : {{ param1 }} <br> {{# output : "defined as var in home" : INCORRECT, should be "defined as param in home"  #}}
param2 : {{ param2 }} <br> {{# output : "defined as param in home" : CORRECT  #}}
param3 : {{ param3 }} <br> {{# output : "" : CORRECT  #}}
{{ param1 = param1 ?? "default value for param1" }}
{{ param2 = param2 ?? "default value for param2" }}
{{ param3 = param3 ?? "default value for param3" }}
param1 : {{ param1 }} <br> {{# output : "defined as var in home" : Still INCORRECT  #}}
param2 : {{ param2 }} <br> {{# output : "defined as param in home" : Still CORRECT  #}}
param3 : {{ param3 }} <br> {{# output : "default value for param3" CORRECT !  #}}
{{ param1 = "modified in partial" }}
<hr>
```

I understand a partial as access by default to the including scope variables, and can alter them, this is Ok, but according to the cascade order doc : https://statamic.dev/data-inheritance#cascade-order
=> explicitly passed parameter should have priority over existing variable in scope

**Comment by JohnathonKoster** [comment]
I'll look into the param1 not being set to the parameter value (which may/may not be a huge breaking change - reminds me a bit of https://github.com/statamic/cms/pull/6405)

However, this behavior is intentional:

```antlers
{{ param1 = "defined as var in home" }}
{{ partial:partials/_test param1="defined as param in home" param2="defined as param in home" }}
home after <br>
param1 : {{ param1 }} <br> {{# output : "modified in partial" : INCORRECT should be "defined as var in home"  #}}
```

You defined the variable before the partial, it was updated in the partial, and that value was pushed up to its defining scope. Changing this would certainly be a massive breaking change.

**Comment by dadaxr** [comment]
I think all params passed to a partial should create their own variable in the partial, untied from the outer scope (even if the param name match an existing outer scope variable). 

However, other existing outer scope variables should still be accessible and be "updatable" from partial. (for what it worth, it's the way it works in twig with the "include" tag).

One more thing, maybe it could be a nice addition to antlers : in twig we can specify if we want outer scope variable to be accessible from included partial or not, with an "only" parameter.

ie : 
![image](https://user-images.githubusercontent.com/1267495/220867524-a68ef632-bbce-41c9-9794-d88c93ef9a5b.png)

**Comment by JohnathonKoster** [comment]
"here param1 beeing passed as a param to the partial should create its own variable in the partial and if updated in the partial, it should not update the "param1" variable from outerscope !"

That would be current behavior before #7588 as all partial parameters can be accessed via. the `view` array (i.e., `{{ view:param1 }}` is isolated from whatever was in the outer scope.

I can see your points on some things, but the current behavior is to allow for things such as this:

```antlers
{{ the_total = 0; }}

{{ loop from="0" to="10" }}
    {{ the_total += 1;
{{ /loop }}
```

The loop tag is just an example, and could be any tag (including the `partial` tag).

**Comment by dadaxr** [comment]
Even if partial and tag pair are somewhat different in philosophy I understand they share similar scoping behavior. Thought In your example, "the_total" is not beeing passed as a parameter of the loop tag, so it should use outer scope variable (which it does).

To stick with your example, IMHO, updating "from" or "to" variables within a loop tag (which does not make sense but let's imagine), should not update any existing "from" or "to" variable in outer scope.

**Comment by dadaxr** [comment]
Also, I wast not aware of the "view" array existence ! Maybe I have missed it from the doc?  If you have a link to share, It would help a lot :)

It's indeed useful to prevent misusing / interfering-with outerscope variables ! thanks for sharing !

**Comment by JohnathonKoster** [comment]
I don't know if that behavior is explicitly documented anywhere. Something I've found while source diving:

That behavior happens on these lines (anything added to `__frontmatter` there will be merged into the `view` array on the frontend):

https://github.com/statamic/cms/blob/3.4/src/Tags/Partial.php#L22-L25

For general documentation around front matter, there is this section: https://statamic.dev/variables/#view-frontmatter

**Comment by dadaxr** [comment]
this is exactly what I was trying to achieve ! Thanks for sharing did not about front matter beeing overrided by passed in parameters ! 

This should be in the doc about how using partials like a pro !

**Comment by dadaxr** [comment]
Ok one more comment to help close this issue and go in the right direction ... 

let's take one more example : 

Considering a "partials/_test" template : 
```
{{# _test partial #}}
---
param1: "default partial frontmatter value"
---
view.param1 before updating : {{ view.param1 }}
{{ view.param1 = "updated value in partial" }}
view.param1 after updating : {{ view.param1 }}

{{ param2 = param2 ? "default partial standard value" }}
param2 before updating  : {{ param2 }}
{{ param2 = "also updated value in partial" }}
param2 after updating  : {{ param2 }}
```

and those independant including templates  : 

- case A
 
```
{{# template A #}}
{{ partial:partials/_test }}
param 1 : {{ param1 }} 
<br>
param 2 : {{ param2 }}  
```

output : 

> == starting partial output ==
> view.param1 before updating : default partial frontmatter value
> 
> view.param1 after updating : updated value in partial
> 
> param2 before updating  : default partial standard value
> 
> param2 after updating  : also updated value in partial
> == finishing partial output ==
> param 1 : 
> param 2 : 

=> this is all good ! (no leak, and correct assignment)



- case B

```
{{# template B #}}
{{ partial:partials/_test param1="defined as param raw value" param2="also defined as param raw value" }}

param 1 : {{ param1 }}
<br>
param 2 : {{ param2 }}
```

output : 

> == starting partial output ==
> view.param1 before updating : defined as param raw value
> 
> view.param1 after updating : updated value in partial
> 
> param2 before updating  : also defined as param raw value
> 
> param2 after updating  : also updated value in partial
> == finishing partial output ==
> param 1 :
> param 2 :

=> this is all good ! (no leak, and correct assignment)



- case C

```
{{# template C #}}
{{ param1 = "defined as var in including template" }}
{{ param2 = "also defined as var in including template" }}
{{ partial:partials/_test param1="defined as param raw value" param2="also defined as param raw value" }}

param 1 : {{ param1 }}
<br>
param 2 : {{ param2 }}
```

output : 

> == starting partial output ==
> view.param1 before updating : defined as param raw value
> 
> view.param1 after updating : updated value in partial
> 
> param2 before updating  : **also defined as var in including template**
> 
> param2 after updating  : also updated value in partial
> == finishing partial output ==
> param 1 : defined as var in including template
> param 2 : **also updated value in partial**

=> here, bold values are wrong : 
> param2 before updating  : **also defined as var in including template**
=> should be : "also defined as param raw value"
(the PR #7588 try to fix this BUT not the right way, as it update the param1 var value instead of injecting an independant value)

> param 2 : **also updated value in partial**
=> should be : "also defined as var in including template" (it should not leak!)
(with PR #7588 merged, the value would still be **also updated value in partial**) 


- case D

```
{{# template D #}}
{{ param1 = "defined as var in including template" }}
{{ param2 = "also defined as var in including template" }}
{{ partial:partials/_test }}

param 1 : {{ param1 }}
<br>
param 2 : {{ param2 }}
```

output : 

> == starting partial output ==
> view.param1 before updating : default partial frontmatter value
> 
> view.param1 after updating : updated value in partial
> 
> param2 before updating  : also defined as var in including template
> 
> param2 after updating  : also updated value in partial
> == finishing partial output ==
> param 1 : defined as var in including template
> param 2 : also updated value in partial

=> this is all good (param2 should indeed be updated in both scope)


- case E

```
{{# template E #}}
{{ param1 = "defined as var in including template but passed by ref" }}
{{ param2 = "also defined as var in including template but passed by ref" }}
{{ partial:partials/_test :param1="param1" :param2="param2" }}

param 1 : {{ param1 }}
<br>
param 2 : {{ param2 }}

```

output : 

> == starting partial output ==
> view.param1 before updating : defined as var in including template but passed by ref
> 
> view.param1 after updating : updated value in partial
> 
> param2 before updating  : also defined as var in including template but passed by ref
> 
> param2 after updating  : also updated value in partial
> == finishing partial output ==
> param 1 : defined as var in including template but passed by ref
> param 2 : also updated value in partial

=> this is all good !  (param2 should indeed be updated in both scope)


TO sum up : 
when a parameter is passed explicitly to a partial we can distinguish 2 cases : 
- if the parameter is passed by reference (:param="xxx") : 
  - it should indeed "leak" (ie update from partial propagate to outer scope, because ... it's the same variable)
- else (the parameter is passed by value copy) : 
  - it should NOT "leak" (it's not the same variable)

@JohnathonKoster do you think your PR #7588 can be updated to reflect expected behavior (see case C) ?

In addition, as it's safer to use frontmatter and "view" variable  when dealing with partial parameters, I suggest to update the "partial related" doc : https://statamic.dev/tags/partial#passing-data to add a recommended way to deal with params. Glad to help with a PR to update the doc if you want to, let me know

**Comment by JohnathonKoster** [comment]
I will be deferring judgement on this matter to the Core team, as I am not comfortable making dramatic changes to scoping/parameter behavior at the risk of introducing massive breaking changes to everyone already using things.

**Comment by dadaxr** [comment]
thanks @duncanmcclean for following up (not the easiest issue to read)

I guess it's OK to close it ... but when you encounter this "weird" behavior and your not familiar with the "inner workings" you go crazy quite fast and the "trust" you put in statamic is somehow altered...

The thing is you can encounter it quite fast when you try to make reusable partials with default parameter values inside, or try to do advanced things in the "wrong way" (without knowing it's wrong ;) ie : https://github.com/statamic/cms/issues/8175) 

But as previously told, this problem is easily mitigated when using the "view" variable inside partials.  I really think this "best" practice (ie: (see:  https://github.com/statamic/cms/issues/7587#issuecomment-1441852921)) should be documented somewhere. 2 spots I suggest : 

1. https://statamic.dev/variables/#view-frontmatter
2. https://statamic.dev/tags/partial#passing-data

I can open a PR on the "Doc" if you want, to try to explain how to use front matter for handling default value.

**Comment by duncanmcclean** [comment]
Feel free to open a PR on the `docs` repository with any changes you think should be made.
